Adds parser and Semantic support for Interop Construct, Init and Use clauses.

swatheesh-mcw opened the pull request:
Adds Parser and Semantic Support for the below construct and clauses:
- Interop Construct
- Init Clause
- Use Clause

Thirumalai-Shaktivel commented:
Yes, I can see it. I will review it later today.

swatheesh-mcw commented:
> Yes, I can see it. I will review it later today.

Sure, thanks!

Thirumalai-Shaktivel commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3738,6 +3738,29 @@ struct OmpOrderClause {
+// struct InteropPreference {
+//   UNION_CLASS_BOILERPLATE(InteropPreference);
+//   std::variant<Name,IntConstantExpr> u;
+// };
```
For string literals, I think we can't use name here, Name is only used for storing Identifiers like variable name.

Use something like CharExpr or similar.

Thirumalai-Shaktivel commented on flang/lib/Parser/openmp-parsers.cpp:
```diff
@@ -473,6 +473,19 @@ TYPE_CONTEXT_PARSER("Omp LINEAR clause"_en_US,
+              "TARGET" >> pure(InteropType::Kind::Target)))
+
+// InteropPreference
+// TYPE_PARSER(construct<InteropPreference>(name || intConstantExpr))
```
Check out some definitions in `expr-parsers.h`. And choose whichever suites here.

swatheesh-mcw replied:
**I tried the following code:**
`TYPE_PARSER(construct<InteropPreference>(construct<InteropPreference>(defaultCharExpr) || 
    construct<InteropPreference>(intConstantExpr)))`

But it always expects **Character** expression to be present rather than allowing either **Character** Expression or **Integer Constant** Expression inside the `InteropPreference`. But similar code is used for `OmpUpdateClause` and there it expects any one of the multiple type to be present. Is there something I am missing for adding the parser support for `InteropPreference`?

Thirumalai-Shaktivel replied:
Not sure, try to debug and fix as much as you can. I will check it out tomorrow, if possible.

swatheesh-mcw replied:
I have tried making couple of modifications but still the issue persist. If you have resolved this, please let me know. Thanks!

Thirumalai-Shaktivel reviewed (commented):
> Also for the InteropPreferenceList, I have currently added support for Integer Constant Datatype. But in the documentation, it is mentioned to also have String Literal. When I tried adding the string literal support, I faced with issues while adding the parser support. I have commented my implementation for InteropPreference in both parse-tree.h and openmp-parsers.cpp files. Can you please have a look into it and let me know how I can resolve that issue? Thanks in advance!

.

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+            if(ObjectNameList.end() != std::find(ObjectNameList.begin(),ObjectNameList.end(),ObjectName)) {
+              context_.Say(GetContext().clauseSource,
+                          "Each interop-var may be speciﬁed for at most one action-clause "
+                          "of each interop construct."_err_en_US);
```
Let's add a test for Semantic checks as well.

swatheesh-mcw replied:
Added tests for semantic checks on [#a36032d](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/a36032d4920bdb75940a3f29597c195893396b0c)

Thirumalai-Shaktivel commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -2090,6 +2090,18 @@ class UnparseVisitor {
     Walk(std::get<OmpObjectList>(x.t));
   }
+  void Unparse(const OmpInitClause::InteropTypes &x) {
+    Walk(x.v,",");
```
```suggestion
    Walk(x.v, ",");
```

swatheesh-mcw replied:
Addressed the review comment on [#a36032d](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/a36032d4920bdb75940a3f29597c195893396b0c).

Thirumalai-Shaktivel commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -2090,6 +2090,18 @@ class UnparseVisitor {
+    Walk(x.v,",");
+  }
+  void Unparse(const OmpInitClause::InteropPreferenceList &x) {
+    Walk(x.v,",");
```
```suggestion
    Walk(x.v, ",");
```

swatheesh-mcw replied:
Addressed the review comment on [#a36032d](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/a36032d4920bdb75940a3f29597c195893396b0c).

Thirumalai-Shaktivel commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -2090,6 +2090,18 @@ class UnparseVisitor {
+    Walk(x.v,",");
+  }
+  void Unparse(const OmpInitClause &x) {
+    Walk("PREFER_TYPE(",std::get<std::optional<OmpInitClause::InteropPreferenceList>>(x.t),"),");
```
```suggestion
    Walk("PREFER_TYPE(", std::get<std::optional<OmpInitClause::InteropPreferenceList>>(x.t), "),");
```

swatheesh-mcw replied:
Addressed the review comment on [#a36032d](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/a36032d4920bdb75940a3f29597c195893396b0c).

Thirumalai-Shaktivel commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -2090,6 +2090,18 @@ class UnparseVisitor {
+  void Unparse(const OmpInitClause &x) {
+    Walk("PREFER_TYPE(",std::get<std::optional<OmpInitClause::InteropPreferenceList>>(x.t),"),");
+    Walk(std::get<OmpInitClause::InteropTypes>(x.t));
+    Put(":");
```
```suggestion
    Put(": ");
```

swatheesh-mcw replied:
Addressed the review comment on [#a36032d](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/a36032d4920bdb75940a3f29597c195893396b0c).

Thirumalai-Shaktivel reviewed (commented):
NIT

Thirumalai-Shaktivel commented:
Very good, @swatheesh-mcw. Sorry, I didn't had much time to checkout the issue.

Anyhow, you fixed it. I hope this PR is ready for review.
I will do a final review, tomorrow.

swatheesh-mcw commented:
> Very good, @swatheesh-mcw. Sorry, I didn't had much time to checkout the issue.
> 
> Anyhow, you fixed it. I hope this PR is ready for review. I will do a final review, tomorrow.

Thanks @Thirumalai-Shaktivel. I got some help in fixing this issue. 

The PR is ready for the review. Please let me know if any changes are required.

kiranktp commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3738,6 +3738,29 @@ struct OmpOrderClause {
   std::tuple<std::optional<OmpOrderModifier>, Type> t;
 };
 
+// InteropType
```
Add the standard ref numbers:
Eg: // Ref: [4.5:169-170], [5.0:255-256], [5.1:288-289], [5.2:323-324] 
for depend clause.

swatheesh-mcw replied:
Sure, will add it.

swatheesh-mcw replied:
Addressed the review comment on [#8d8dc3a](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/8d8dc3ac1dbfb6bc92802da7bd6504513ad60692)

kiranktp commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3738,6 +3738,33 @@ struct OmpOrderClause {
   std::tuple<std::optional<OmpOrderModifier>, Type> t;
 };
 
+// REF: [5.1:217-220], [5.2:293-294]
```
let us keep all changes related to interop at one place in this file. currently directive definition is in different place and the init clause def is in different place.

swatheesh-mcw replied:
Hi Kiran, Since the Interop construct uses clauselists in its arguments, its definition should follow after clauselist definition which is after line 3872. The clauses should be defined before line 3856 as all the mentioned clauses will get generated in llvm/Frontend/OpenMP/OMP.inc file. So, we cannot define both interop construct and init clauses definitions together. Hence, I implemented the respective definitions in different places.

kiranktp commented on flang/include/flang/Parser/dump-parse-tree.h:
```diff
@@ -581,6 +581,14 @@ class ParseTreeDumper {
+  NODE(parser, InteropType)
+  NODE_ENUM(InteropType, Kind)
+  NODE(parser, InteropPreference)
+  NODE(OmpInitClause, InteropPreferenceList)
```
Better to replace InteropPreferenceList with InteropModifier to be in sync with the standard.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

kiranktp commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3738,6 +3738,33 @@ struct OmpOrderClause {
+
+struct OmpInitClause {
+  TUPLE_CLASS_BOILERPLATE(OmpInitClause);
+  WRAPPER_CLASS(InteropPreferenceList, std::list<InteropPreference>);
```
Better to replace InteropPreferenceList with InteropModifier to be in sync with the standard.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

kiranktp commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3738,6 +3738,33 @@ struct OmpOrderClause {
+// REF: [5.1:217-220], [5.2:293-294]
+//
+// init-clause -> INIT ([interop-modifier,] [interop-type,]
+//                              interop-type: interop-var)
```
expand interop-modifier and interop-type as well in the comment section:
		interop-modifier: prefer_type(preference-list)
		interop-type: target, targetsync
			There can be at most only two interop-type.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

kiranktp commented on flang/lib/Parser/openmp-parsers.cpp:
```diff
@@ -473,6 +473,22 @@ TYPE_CONTEXT_PARSER("Omp LINEAR clause"_en_US,
+TYPE_PARSER(construct<InteropType>(
+              "TARGETSYNC" >> pure(InteropType::Kind::TargetSync) ||
+              "TARGET" >> pure(InteropType::Kind::Target)))
+
```
Remove extra line

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

kiranktp reviewed (commented):
LGTM

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+            } else {
+              ++targetCount;
+            }
+            if (targetCount >= 2 || targetSyncCount >= 2) {
```
NIT:
This seems confusing, how about?
```suggestion
            if (targetCount > 1 || targetSyncCount > 1) {
```

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+    },
+    clause.u);
+  }
+  if (isDependClauseOccured && !isTargetSyncOccured) {
```
Let's check `targetSyncCount` is `0`, instead of using a new variable `isTargetSyncOccured`
```suggestion
  if (isDependClauseOccured && !targetSyncCount) {
```

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+          const auto &InteropVar{parser::Unwrap<parser::OmpObject>(std::get<parser::OmpInitClause::InteropVar>(InitClause.v.t))};
+          const auto *name{parser::Unwrap<parser::Name>(InteropVar)};
+          const auto ObjectName{name->ToString()};
+          if(ObjectNameList.empty()) {
+            ObjectNameList.push_back(ObjectName);
+          } else {
+            if(ObjectNameList.end() != std::find(ObjectNameList.begin(),ObjectNameList.end(),ObjectName)) {
+              context_.Say(GetContext().clauseSource,
+                          "Each interop-var may be speciﬁed for at most one action-clause "
+                          "of each interop construct."_err_en_US);
+            }
+          }
```
```suggestion
          if(ObjectNameList.end() != std::find(ObjectNameList.begin(),ObjectNameList.end(),ObjectName)) {
            context_.Say(GetContext().clauseSource,
                        "Each interop-var may be speciﬁed for at most one action-clause "
                        "of each interop construct."_err_en_US);
          } else {
            ObjectNameList.push_back(ObjectName);
          }
```

Thirumalai-Shaktivel replied:
Or you can use `SmallVectorImpl`. This simplifies the check.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel replied:
Usually, it is recommended to use `SmallVectorImpl` or `ArrayRef` here instead of `std::list`

These containers has some useful builtin functions. Like there is something like `contained` function to check for element in the list

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+          const auto &InteropVar{parser::Unwrap<parser::OmpObject>(DestroyClause.v)};
+          const auto *name{parser::Unwrap<parser::Name>(InteropVar)};
+          const auto ObjectName{name->ToString()};
+          if(ObjectNameList.empty()) {
```
Do the same here.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+          const auto &InteropVar{parser::Unwrap<parser::OmpObject>(UseClause.v)};
+          const auto *name{parser::Unwrap<parser::Name>(InteropVar)};
+          const auto ObjectName{name->ToString()};
+          if(ObjectNameList.empty()) {
```
Do the same here.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+            }
+          }
+        },
+        [&](const auto &) {},
```
We visit all the clauses already, do we still need this?

Thirumalai-Shaktivel replied:
I see, we don't visit device, and nowait.

So, it's fine.

Thirumalai-Shaktivel commented on flang/test/Parser/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,204 @@
+  use omp_lib
+  integer(omp_interop_kind) :: obj
+  integer, dimension(1,10) :: arr
+  !$omp interop init(prefer_type("cuda"),targetsync,target: obj) depend(inout: arr) nowait
```
Is the following case valid and does it work in this PR?
```fortran
!$omp interop init(prefer_type("cu"//"da"): obj)
```

swatheesh-mcw replied:
I am not sure if the above mentioned case is valid, but it won't work with this PR.

Thirumalai-Shaktivel reviewed (commented):
Overall, looks good. Left some minor comments

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,95 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+                          "at most once."_err_en_US);
+            } 
+          }
+          const auto &InteropVar{parser::Unwrap<parser::OmpObject>(std::get<parser::OmpInitClause::InteropVar>(InitClause.v.t))};
```
Run 
```console
$ git-clang-format
```

This might fail while submitting an upstream PR.

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,25 @@
+!RUN: not %flang_fc1 -fopenmp %s 2>&1 | FileCheck %s
```
Why `not`, is there any failure?

If any problem, do let me know.

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,25 @@
+!RUN: not %flang_fc1 -fopenmp %s 2>&1 | FileCheck %s
+
```
Add which OpenMP version and what you are doing in this test.

See other tests for reference in the same folder

swatheesh-mcw replied:
I have taken reference from llvm-project/flang/test/Semantics/OpenMP/invalid-branch.f90 to create this semantic test file.

Thirumalai-Shaktivel replied:
Use this reference: https://github.com/swatheesh-mcw/llvm-project/blob/main/flang/test/Semantics/OpenMP/allocate-clause01.f90

swatheesh-mcw replied:
I tried making my semantic checks similar to the above mentioned example, but for some reason I am not getting the expected semantic check errors . My stdout and stderr is empty and my test case gets failed.

Thirumalai-Shaktivel replied:
Share me the output of
```
$ flang-new -fc1 flang/test/Semantics/OpenMP/interop-construct.f90 -fopenmp
```

swatheesh-mcw replied:
`error: Semantic errors in llvm-project/flang/test/Semantics/OpenMP/interop-construct.f90
error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
error: Each interop-type may be speciﬁed at most once.
error: A depend clause can only appear on the directive if the interop-type includes targetsync`

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,25 @@
+  !CHECK: error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
+  !$omp interop init(targetsync,target: obj) use(obj)
+  print *,'pass'
+end
```
NIT:
```suggestion
end subroutine test_interop_01
```

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,25 @@
+!RUN: not %flang_fc1 -fopenmp %s 2>&1 | FileCheck %s
+
+subroutine main1()
```
```suggestion
subroutine test_interop_01()
```

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,25 @@
+  integer(omp_interop_kind) :: obj
+  !CHECK: error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
+  !$omp interop init(targetsync,target: obj) use(obj)
+  print *,'pass'
```
```suggestion
  print *, 'pass'
```

swatheesh-mcw replied:
Addressed the comment on [#1c57d1e](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/1c57d1ea2b8e7095a834689ad5b832e6a2a84c37)

Thirumalai-Shaktivel reviewed (commented):
NIT: Do the same for other subroutine's

Thirumalai-Shaktivel commented on flang/test/Semantics/OpenMP/interop-construct.f90:
```diff
@@ -0,0 +1,25 @@
+subroutine test_interop_01()
+  use omp_lib
+  integer(omp_interop_kind) :: obj
+  !CHECK: error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
```
```suggestion
  !ERROR: error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
```

swatheesh-mcw replied:
I did check making the above mentioned changes with the test_errors.py file check but for some reason the test_errors.py is not picking up the actual errors from the console log output.

Thirumalai-Shaktivel replied:
Can you share the issue or error you see here?

swatheesh-mcw replied:
If I ran the above code with `! RUN: %python %S/../test_errors.py %s %flang -fopenmp -fopenmp-version=52
` , I am getting the bellow error:


`FAIL: Flang :: Semantics/OpenMP/interop-construct.f90 (1311 of 3042)
******************** TEST 'Flang :: Semantics/OpenMP/interop-construct.f90' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
--- 
+++ 
@@ -0,0 +1,3 @@

expect at 7: error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
expect at 15: error: Each interop-type may be speciﬁed at most once.
expect at 23: error: A depend clause can only appear on the directive if the interop-type includes targetsync

FAIL

--
Command Output (stderr):
--
RUN: at line 1: "/usr/bin/python3.10" /home/mcw/Documents/Swatheesh/llvm-project/flang/test/Semantics/OpenMP/../test_errors.py /home/mcw/Documents/Swatheesh/llvm-project/flang/test/Semantics/OpenMP/interop-construct.f90 /home/mcw/Documents/Swatheesh/build/bin/flang -fopenmp -fopenmp-version=52
+ /usr/bin/python3.10 /home/mcw/Documents/Swatheesh/llvm-project/flang/test/Semantics/OpenMP/../test_errors.py /home/mcw/Documents/Swatheesh/llvm-project/flang/test/Semantics/OpenMP/interop-construct.f90 /home/mcw/Documents/Swatheesh/build/bin/flang -fopenmp -fopenmp-version=52

--

********************
`

Thirumalai-Shaktivel replied:
Tip: GitHub uses Markdown format, you can use three ` to enclose code or output.
See, https://www.markdownguide.org/extended-syntax/#fenced-code-blocks for more features

Thirumalai-Shaktivel replied:
Can you share the output of `flang-new flang/test/Semantics/OpenMP/interop-construct.f90 -fopenmp -fopenmp-version=50` as well?

Thirumalai-Shaktivel replied:
Paste the output in this format
```console
$ flang-new flang/test/Semantics/OpenMP/interop-construct.f90 -fopenmp -fopenmp-version=50
[...]
```

swatheesh-mcw replied:
Thanks for the tip @Thirumalai-Shaktivel , Please see the output mentioned below:

```flang-new llvm-project/flang/test/Semantics/OpenMP/interop-construct.f90 -fopenmp -fopenmp-version=50
flang-20: warning: OpenMP support in flang is still experimental [-Wexperimental-option]
error: Semantic errors in llvm-project/flang/test/Semantics/OpenMP/interop-construct.f90
error: Each interop-var may be speciﬁed for at most one action-clause of each interop construct.
error: Each interop-type may be speciﬁed at most once.
error: A depend clause can only appear on the directive if the interop-type includes targetsync
```

Thirumalai-Shaktivel replied:
Nothing else, is it? 
Doesn't it print the location information of the error?
Something like the following:
```console
$ flang-new x.f90
error: Semantic errors in x.f90
./x.f90:2:24: error: Must be a constant value
      integer :: a(10) = x
                         ^
```

Thirumalai-Shaktivel replied:
If no, then we have to add the source information to the interop construct in the parser.

swatheesh-mcw replied:
Yeah, it doesn't print the location information of the error. 
Can you please let me know where should I add the source information so that I can add it and test whether it is working fine? Thanks!

Thirumalai-Shaktivel replied:
If you are confused, ping me here. I can explain more.

Thirumalai-Shaktivel replied:
See, https://github.com/swatheesh-mcw/llvm-project/pull/1/files#r1886185039

swatheesh-mcw replied:
I am a little confused here, can you just explain a little bit about the source information?

Thirumalai-Shaktivel reviewed (commented):
I see, this is the issue. Do the same for all the other errors as well.

Thirumalai-Shaktivel commented:
@swatheesh-mcw, what is the status here?

Were you able to fix the failure?

Thirumalai-Shaktivel commented:
Also, make sure
- to apply `git clang-format` 
- `ninja check-flang` passes all tests

swatheesh-mcw commented:
> @swatheesh-mcw, what is the status here?
> 
> Were you able to fix the failure?

Hi @Thirumalai-Shaktivel , I could not able to use the test_errors.py file to check for semantic errors. When I tried debugging the issue, I found that the regex used to find the actual error on test_errors.py doesn't seem to pickup semantic errors that are getting printed on the console log. It does pickup the errors that we have mentioned in the unit test. That's why it cannot match the expected errors with actual errors because the actual error list is empty as regex is not picking up those errors from the console output. I could only get the issue solved if I use `!RUN: not %flang_fc1 -fopenmp -fopenmp-version=52 %s 2>&1 | FileCheck %s` for checking semantic errors in the unit test.

swatheesh-mcw commented:
> Also, make sure
> 
>     * to apply `git clang-format`
> 
>     * `ninja check-flang` passes all tests

I have run the above mentioned commands on my latest commit. So, it should fine for the review. Thanks!

Thirumalai-Shaktivel commented:
Adding this is of no use; `!RUN: not %flang_fc1—fopenmp—fopenmp-version=52 %s 2>&1 | FileCheck %s`  doesn't actually check the error. The only way to test your changes is to add `test_error.py ...`

Please share all the issues you are facing we can fix it together and upstream this patch.

Thirumalai-Shaktivel commented on flang/lib/Parser/openmp-parsers.cpp:
```diff
@@ -804,6 +823,10 @@ TYPE_PARSER(sourced(construct<OmpSimpleStandaloneDirective>(first(
     Parser<OmpSimpleStandaloneDirective>{}, Parser<OmpClauseList>{})))
 
+// 14.1 Interop construct
+TYPE_PARSER(sourced(construct<OpenMPInteropConstruct>(
```
Okay, I see you have added `sourced(` here, do the same for the leaf nodes as well.

Thirumalai-Shaktivel commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -4169,6 +4199,15 @@ struct OmpSimpleStandaloneDirective {
+// interop -> INTEROP clause[ [ [,] clause]...]
+struct OpenMPInteropConstruct {
+  TUPLE_CLASS_BOILERPLATE(OpenMPInteropConstruct);
+  CharBlock source;
```
The `source` values here gets updated when you pass the parser result into `sourced` function in `openmp-parsers.cpp`.

So, check if the same is possible for `InteropType`, `InteropPreference` and `OmpInitClause`.

swatheesh-mcw replied:
Hi @Thirumalai-Shaktivel , 

I found the root cause of the issue. The reason why the source information is not carried back to Interop Construct is because we have checked all the semantic checks inside the Interop Construct block on ```check-omp-structures.cpp``` and not on individual clauses. 

So instead of using ```GetContext().clauseSource```, if we use ```GetContext().directiveSource```, we are getting the source information for the error. Hence, the semantic checks are now passing with ```test_errors.py```. 

Addressed the issue on the commit [#9b0cd4b](https://github.com/swatheesh-mcw/llvm-project/pull/1/commits/9b0cd4b2c15449788bb1b66a4832d9c3be8c4195). I forgot to remove ```iostream``` header in this commit so removed that in the next commit.

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,97 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+  int targetCount{0}, targetSyncCount{0};
+  const auto &dir{std::get<parser::Verbatim>(x.t)};
+  std::list<std::string> ObjectNameList;
+  PushContextAndClauseSets(dir.source, llvm::omp::Directive::OMPD_interop);
```
Also, try to check the `source` has some value. Try printing `dir.source` and `GetContext().clauseSource`. Check if they are the same and correct.

The source can be printed as a string.

Thirumalai-Shaktivel reviewed (commented):
.

swatheesh-mcw commented:
Hi @Thirumalai-Shaktivel , I have added the fix for the semantic issue that we were discussing. Please have a look into the commit and let me know if any change is required. Thanks!

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -4550,6 +4550,97 @@ void OmpStructureChecker::Leave(const parser::DoConstruct &x) {
+            }
+            if (targetCount > 1 || targetSyncCount > 1) {
+              context_.Say(GetContext().directiveSource,
+                          "Each interop-type may be speciﬁed "
+                          "at most once."_err_en_US);
```
Recently, Kparzysz suggested keeping the Error message in a single line.
Something like:
```suggestion
                          "Each interop-type may be speciﬁed at most once."_err_en_US);
```
git clang-format won't complain about it.
Do the same for all the other error messages as well.

swatheesh-mcw replied:
Sure, will do it. Thanks!

Thirumalai-Shaktivel reviewed (approved):
LGTM, left minor comment.

I think you can use the same branch to submit a PR for  the llvm-project.
Please make sure it compile and run all the tests successfully.
Finally, apply `git clang-format`

Thirumalai-Shaktivel commented:
Thanks for the update!

swatheesh-mcw committed:
Adds Parser and Semantic Support for the below construct and clauses:

- Interop Construct
- Init Clause
- Use Clause

Note:
The other clauses supported by Interop Construct such as Destroy, Use, Depend and Device are added already.

swatheesh-mcw committed:
Modified the declaration and definition of modifiers and addressed review comments.

swatheesh-mcw committed:
Preserved the alphabetical ordering of modifier declarations and definitions.

swatheesh-mcw committed:
Updated the testfile, rebased and added a test on Todo Lower side for interop construct.

swatheesh-mcw committed:
Addressed review comments and changed object name check with object symbol check.

swatheesh-mcw committed:
Addressed NIT comments.

swatheesh-mcw committed:
Changed the InteropRuntimeIdentifier comment on parse-tree.h

swatheesh-mcw committed:
Formatted check-omp-structures.cpp

swatheesh-mcw committed:
Formatted using git clang-format

swatheesh-mcw committed:
Modified the testcases for Interop construct
